cmd/dump.go

```diff
@@ -0,0 +1,285 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	gatewayapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+	"sigs.k8s.io/yaml"
+
+	kuadrantv1alpha1 "github.com/kuadrant/kuadrant-operator/api/v1alpha1"
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+	kuadrantv1beta2 "github.com/kuadrant/kuadrant-operator/api/v1beta2"
+)
+
+var (
+	dumpNamespace     string
+	dumpAllNamespaces bool
+	dumpOutputDir     string
+	dumpFormat        string
+)
+
+func dumpCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "dump",
+		Short: "Dump Kuadrant and Gateway API resources",
+		Long:  "Dump all Kuadrant and Gateway API resources to files for investigation and debugging",
+		RunE:  runDump,
+	}
+
+	cmd.Flags().StringVarP(&dumpNamespace, "namespace", "n", "", "Namespace to dump resources from (default: all namespaces)")
+	cmd.Flags().BoolVarP(&dumpAllNamespaces, "all-namespaces", "A", false, "Dump resources from all namespaces")
+	cmd.Flags().StringVarP(&dumpOutputDir, "output", "o", "", "Output directory (default: ./kuadrant-dump-<timestamp>)")
+	cmd.Flags().StringVar(&dumpFormat, "format", "yaml", "Output format: yaml or json")
+
+	return cmd
+}
+
+func runDump(cmd *cobra.Command, _ []string) error {
+	ctx := cmd.Context()
+	logger := logf.FromContext(ctx)
+
+	// Determine output directory
+	if dumpOutputDir == "" {
+		timestamp := time.Now().Format("20060102-150405")
+		dumpOutputDir = fmt.Sprintf("kuadrant-dump-%s", timestamp)
+	}
+
+	// Create output directory
+	if err := os.MkdirAll(dumpOutputDir, 0755); err != nil {
+		return fmt.Errorf("failed to create output directory: %w", err)
```

Comment:
Can we only create the directory when we are ready to save data. There is a number of failure points between directory creation and requiring the directory to exist.